prov/util/src/util_cntr.c

```diff
@@ -0,0 +1,317 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+int ofi_check_cntr_attr(const struct fi_provider *prov,
+			const struct fi_cntr_attr *attr)
+{
+	if (!attr)
+		return FI_SUCCESS;
+
+        if (attr->flags) {
+		FI_WARN(prov, FI_LOG_CQ, "unsupported flags\n");
+		return -FI_EINVAL;
+	}
+
+	switch (attr->wait_obj) {
+	case FI_WAIT_NONE:
+		break;
+	case FI_WAIT_SET:
+		if (!attr->wait_set) {
+			FI_WARN(prov, FI_LOG_CNTR, "invalid wait set\n");
+			return -FI_EINVAL;
+		}
+		/* fall through */
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		break;
+	default:
+		FI_WARN(prov, FI_LOG_CNTR, "unsupported wait object\n");
+		return -FI_EINVAL;
+	}
+
+	return 0;
+}
+
+static uint64_t ofi_cntr_read(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+	return ofi_atomic_get64(&cntr->cnt);
+}
+
+static uint64_t ofi_cntr_readerr(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+	return ofi_atomic_get64(&cntr->err);
+}
+
+static int ofi_cntr_add(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->cnt, value);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_adderr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->err, value);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_set(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->cnt, value);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_seterr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass = FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->err, value);
+
+	return FI_SUCCESS;
+}
```

Comment:
Please add TODO to add/adderr/set/seterr to signal the wait object if the counter values have crossed some given threshold.  You may want to take a look at the psm provider implementation of counters.  I *think* it handles this.